### DISCUSSION TITLE
What is Neovim's frame rate and tick rate?

### DISCUSSION THREAD

**Discussion by ghost** [neovim/neovim#25570] [Q&A] [answered]
I can't find this in the code.

**Comment by zeertzjq** [comment]
There is no such thing because Nvim doesn't have a builtin GUI.

**Reply by ghost** [reply]
I mean it in the sense of frame rate is how often the screen is rendered to and tick rate is how often input events are read. So for example delaying 4ms, reading/handling input events, repeat.

**Reply by zeertzjq** [reply, marked as answer]
Polling for input events is handled by libuv, apart from that there is no delay at all. The screen is updated immediately when something needs change.

**Reply by ghost** [reply]
So wouldn't that be synchronous, not asynchronous? Wait for input -> render with new state once input happens -> repeat

**Reply by zeertzjq** [reply]
Yes.

**Reply by ghost** [reply]
I'm confused, I thought libuv was an asynchronous library? Can you explain how this all works? Does it listen for input in a new thread or something? Thanks! Unfortunately I don't know C, I originally asked this question because of [this Rust TUI library tutorial](https://ratatui.rs/tutorial/counter-async-app/async-event-stream.html), and I'm wondering about how these text editors are built efficiently.

**Reply by dundargoc** [reply]
> efficiently

You must be new here ;)

**Reply by ghost** [reply]
It seems pretty efficient to me.

**Reply by justinmk** [reply]
> Does it listen for input in a new thread or something?

A separate thread isn't necessary. libuv abstracts OS mechanisms that asynchronously invoke Nvim's main loop when the OS receives input.
